src/stream/src/executor/aggregation/mod.rs

```diff
@@ -342,6 +342,59 @@ pub fn generate_agg_schema(
     Schema { fields }
 }
 
+/// Infer column desc for state table.
+/// The column desc layout is
+/// [ `group_key` (only for hash agg) / `sort_key` (only for simple agg) / `value`(the agg call
+/// return type)].
+/// This is the Row layout insert into state table.
+/// For different agg call, different executor (hash agg or simple agg), the layout will be
+/// different.
+pub fn generate_column_descs(
+    agg_call: &AggCall,
+    group_keys: &[usize],
+    pk_indices: &[usize],
+    agg_schema: &Schema,
+    input_ref: &dyn Executor,
+) -> Vec<ColumnDesc> {
+    let mut column_descs = Vec::with_capacity(group_keys.len() + 1);
+    let mut column_id = 0;
+    for (idx, _) in group_keys.iter().enumerate() {
+        // The agg schema layout as [group_key schema, agg_call schema], so directly get group key
+        // return type from it for DRY.
+        column_descs.push(ColumnDesc::unnamed(
+            ColumnId::new(column_id),
+            agg_schema.fields[idx].data_type.clone(),
+        ));
+        column_id += 1;
```

Comment:
Suggest introducing a closure of `next_column_id` here. :)

Reply:
Great! The code looks more elegant